File: Graduation_Project_BE/src/main/java/com/project/app/report/service/ReportService.java
package com.project.app.report.service;

import com.project.app.report.dto.response.ReportDistributionResponse;
import com.project.app.report.dto.response.ReportTrendResponse;
import com.project.app.transaction.entity.TransactionType;
import com.project.app.user.entity.User;

import java.time.LocalDate;
import java.util.List;

public interface ReportService {
    List<ReportDistributionResponse> getDistributionReport(User user, TransactionType type, String filter, LocalDate date);
    
    List<ReportTrendResponse> getTrendReport(User user, TransactionType type, String filter, LocalDate date);
}

File: Graduation_Project_FE/shared/api/services/reportService.ts
import { axiosClient } from '../axiosClient';
import { ENDPOINTS } from '../endpoints';

export interface ReportDistributionResponse {
    categoryId: number | null;
    categoryName: string;
    icon: string;
    color: string;
    totalAmount: number;
    percentage: number;
}

export interface ReportTrendResponse {
    label: string;
    value: number;
    isCurrent: boolean;
}

export const reportService = {
    getDistributionReport: async (type: 'EXPENSE' | 'INCOME', filter: string, date: string): Promise<ReportDistributionResponse[]> => {
        try {
            const response: any = await axiosClient.get(ENDPOINTS.REPORT.DISTRIBUTION, {
                params: { type, filter, date }
            });
            return response.data || [];
        } catch (error) {
            console.error("Error fetching distribution report:", error);
            return [];
        }
    },

    getGroupDistributionReport: async (type: 'EXPENSE' | 'INCOME', filter: string, date: string): Promise<ReportDistributionResponse[]> => {
        try {
            const response: any = await axiosClient.get(ENDPOINTS.REPORT.DISTRIBUTION, {
                params: { type, filter, date, groupBy: 'GROUP' }
            });
            return response.data || [];
        } catch (error) {
            console.error("Error fetching group distribution report:", error);
            return [];
        }
    },

    getTrendReport: async (type: 'EXPENSE' | 'INCOME', filter: string, date: string): Promise<ReportTrendResponse[]> => {
        try {
            const response: any = await axiosClient.get(ENDPOINTS.REPORT.TREND, {
                params: { type, filter, date }
            });
            return response.data || [];
        } catch (error) {
            console.error("Error fetching trend report:", error);
            return [];
        }
    }
};
